mobile-browser-based-version/server/run_server.js

```diff
@@ -1,86 +1,293 @@
+    const weightsPath = `${task}_${round}_weights.json`;
+    const milestonesPath = path.join(__dirname, 'milestones');
+    if (!fs.existsSync(milestonesPath)) {
+      fs.mkdirSync(milestonesPath);
+    }
+    fs.writeFile(path.join(milestonesPath, weightsPath), weightsJson, (err) => {
+      if (err) {
+        console.log(err);
+        console.log(`Failed to save weights to ${weightsPath}`);
+      } else {
+        console.log(`Weights saved to ${weightsPath}`);
+      }
+    });
+  }
+
+  let weights = msgpack.encode(Array.from(serializedWeights));
+  response.status(200).send({weights: weights});
+  return;
 }
 
-Promise.all(models.map((createModel) => createModel()))
+function sendDataInfo(request, response) {
+  const requestType = 'SEND_data_info';
 
+  if (!isValidRequest(request)) {
+    response.status(400).send(INVALID_REQUEST_FORMAT_MESSAGE);
+    console.log(`${requestType} failed`);
+    return;
+  }
+
+  const id = request.body.id;
+  const timestamp = request.body.timestamp;
+  const samples = request.body.samples;
+  const task = request.params.task;
+  const round = request.params.round;
+
+  if (!(task in dataDict)) {
+    dataDict[task] = {};
+  }
+  if (!(round in dataDict)) {
+    dataDict[task][round] = {};
+  }
+  dataDict[task][round][id] = samples;
+  response.status(200).send("Number of samples successfully received.");
+
+  logsAppend(request, requestType);
+  return;
+}
+
+function receiveDataInfo(request, response) {
+  const requestType = 'RECEIVE_data_info';
+
+  if (!isValidRequest(request)) {
+    response.status(400).send(INVALID_REQUEST_FORMAT_MESSAGE);
+    console.log(`${requestType} failed`);
+    return;
+  }
+
+  const body = request.body;
+  const id = body.id;
+  const timestamp = body.timestamp;
+  const task = request.params.task;
+  const round = request.params.round;
+
+  if (!(task in dataDict && round >= 0)) {
+    response.status(404).send(INVALID_REQUEST_KEYS_MESSAGE);
+    console.log(`${requestType} failed`);
+    return;
+  }
+
+  const latestDataDict = {};
+  // For each round...
+  for (let data of Object.values(dataDict[task])) {
+    // ... fetch the latest entry
+    for (let [id, samples] of Object.entries(data)) {
+      latestDataDict[id] = samples;
+    }
+  }
+
+  /* Code for computing data percentages, left to client
+  const totalSamples = Object.values(latestDataDict).reduce((a, b) => {
+    a + b
+  });
+
+  // Map values
+  const dataShares = {};
+  for (let [id, samples] of Object.entries(latestDataDict)) {
+    dataShares[id] = samples / totalSamples;
+  }
+  */
+
+  response.status(200).send(latestDataDict);
+
+  logsAppend(request, requestType);
+  return;
+}
+
+function getTasks(req, res) {
+  const tasksPath = path.join(__dirname, 'tasks.json');
+  if (fs.existsSync(tasksPath)) {
+    console.log(`Serving ${tasksPath}`);
+    res.status(200).sendFile(tasksPath);
+  } else {
+    res.status(400).send({});
+  }
+}
+
+function getTaskModel(req, res) {
```

Comment:
which model is this. initial one or current one? from server or client? change the function name to be more expessive and add a comment string